email_account_validity/_store.py

```diff
@@ -246,8 +247,36 @@ def get_expiration_ts_for_user_txn(txn: LoggingTransaction):
         )
         return res
 
-    async def set_renewal_token_for_user(self, user_id: str, renewal_token: str):
+    async def set_renewal_token_for_user(
+        self,
+        user_id: str,
+        renewal_token: str,
+        unique: bool,
+    ):
+        """Store the given renewal token for the given user.
+
+        Args:
+            user_id: The user ID to store the renewal token for.
+            renewal_token: The renewal token to store for the user.
+            unique: Whether the token should be unique across the whole database, i.e.
+                whether it should be able to look the user up from the token.
+
+        Raises:
+            SynapseError(409): unique is set to True and the token is already in use.
+        """
         def set_renewal_token_for_user_txn(txn: LoggingTransaction):
+            if unique:
+                ret = DatabasePool.simple_select_one_onecol_txn(
+                    txn=txn,
+                    table="email_account_validity",
+                    keyvalues={"renewal_token": renewal_token},
+                    retcol="user_id",
+                    allow_none=True,
+                )
+
+                if ret is not None:
+                    raise SynapseError(409, "Renewal token already in use")
```

Comment:
could we not replace this with an `insert` (assuming we have the right index on the table)?

Reply:
We can't really have an index because we don't want to enforce a unique constraint on all of the tokens: if it's a small one used with authentication we don't want it to be unique across the whole table (otherwise we risk running into situations where we're not able to generate them because we hit 5+ clashes when trying to find one that isn't already used).

Though maybe we should be storing both types of tokens in different (nullable) columns, which would allow having a unique index on them, wdyt?

Reply:
> Though maybe we should be storing both types of tokens in different (nullable) columns, which would allow having a unique index on them, but sounds like a slightly more confusing schema to me, wdyt?

I'm actually going to do this, since not having an index on this sounds like a bad thing.